src/host-services.js

```diff
@@ -0,0 +1,150 @@
+   */
+  static visibilityForDoc(elementOrAmpDoc) {
+    return /** @type {!Promise<!Visibility>} */ (
+      getServicePromiseForDoc(elementOrAmpDoc, ServiceNames.VISIBILITY));
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @param {!Visibility} impl
+   */
+  static installVisibilityServiceForDoc(elementOrAmpDoc, impl) {
+    registerServiceBuilderForDoc(elementOrAmpDoc,
+        ServiceNames.VISIBILITY, impl, /* opt_instantiate */ true);
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @return {!Promise<!Visibility>}
+   */
+  static fullscreenForDoc(elementOrAmpDoc) {
+    return /** @type {!Promise<!Fullscreen>} */ (
+      getServicePromiseForDoc(elementOrAmpDoc, ServiceNames.FULLSCREEN));
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @param {!Fullscreen} impl
+   */
+  static installFullscreenServiceForDoc(elementOrAmpDoc, impl) {
+    registerServiceBuilderForDoc(elementOrAmpDoc,
+        ServiceNames.FULLSCREEN, impl, /* opt_instantiate */ true);
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @return {!Promise<!Exit>}
+   */
+  static exitForDoc(elementOrAmpDoc) {
+    return /** @type {!Promise<!Fullscreen>} */ (
+      getServicePromiseForDoc(elementOrAmpDoc, ServiceNames.EXIT));
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @param {!Exit} impl
+   */
+  static installExitServiceForDoc(elementOrAmpDoc, impl) {
+    registerServiceBuilderForDoc(elementOrAmpDoc,
+        ServiceNames.EXIT, impl, /* opt_instantiate */ true);
+  }
+}
+
+/**
+ * Visibility defines interface provided by host for visibility detection.
+ *
+ * @interface
+ */
+export class Visibility {
+
+  /**
+   * Register a callback for scroll events.
+   * @param {function(!VisibilityDataDef)} unusedCallback
+   */
+  onVisibilityChange(unusedCallback) {}
+}
+
+/**
+ * The structure that combines position and size for an element. The exact
+ * interpretation of position and size depends on the use case.
+ *
+ * @typedef {{
+ *   visibleRect: (?./layout-rect.LayoutRectDef),
+ *   visibleRatio: number
+ * }}
+ */
+export let VisibilityDataDef;
+
+
+/**
+ * Fullscreen defines interface provided by host to enable/disable
+ * fullscreen mode.
+ *
+ * @interface
+ */
+export class Fullscreen {
+
+  /**
+   * Request to expand the given element to fullscreen overlay.
+   *
+   * @param {!Element} unusedTargetElement
+   * @return {!Promise<boolean>}
+   */
+  enterFullscreen(unusedTargetElement) {}
+
+  /**
+   * Request to exit from fullscreen overlay.
+   *
+   * @param {!Element} unusedTargetElement
+   * @return {!Promise<boolean>}
+   */
+  exitFullscreen(unusedTargetElement) {}
+}
+
+/**
+ * Exit defines interface provided by host for navigating out.
+ *
+ * @interface
+ */
+export class Exit {
+
+  /**
+   * Request to navigate to URL.
```

Comment:
shall there be an option for new tab vs top navigation (like target="_blank")? Or will AMP always prefer new tabs?

Reply:
I saw amp-ad-exit only do target="_blank". Does it support other target type?

Reply:
No. I guess it'd be easy to add later if needed.